### PR TITLE
[Spring] Add apiFirst option

### PR DESCRIPTION
**Pull request by cbornet**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the correct branch: Default: `master`.
- [x] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

This PR:
* adds a `apiFirst` option that will not generate the API but will configure the openapi-generator-maven-plugin for API-first developpement instead.
* removes usage of Springfox and serves the OAI spec directly if `apiFirst` or `reactive` is set (as Springfox doesn't support webflux yet)
* removes some generated classes that have no use in Spring (ApiException, ApiOriginFilter, ApiResponseMessage, NotFoundException)
* fixes configuration of ThreetenBP for Spring-MVC
* configures the CORS globally (code to uncomment)

### CODE REVIEW

**Comment by cbornet** [comment]
cc @bbdouglas (2017/07) @JFCote (2017/08) @sreeshas (2017/08) @jfiala (2017/08) @lukoyanov (2017/09) @cbornet (2017/09) @jeff9finger (2018/01)

**Comment by cbornet** [comment]
@jmini No I'm sure I updated the samples. Can you have a look on your side ? I suspect code generated might be different depending on the machine it runs on.

**Comment by jmini** [comment]
Same here. I do not have any changes.

```
git fetch origin pull/184/head:pull_184
git checkout pull_184
mvn clean verify
bin/springboot-petstore-server-reactive.sh
git status
```

Gives me this output:
```
On branch pull_184
nothing to commit, working tree clean
```

---
For investigation, can you maybe add a line with `git diff` after `git status` in the script:
https://github.com/OpenAPITools/openapi-generator/blob/7dfd940023a717649aa9af22100690ddab2731c1/bin/ensure-up-to-date#L29-L34

This way we will see what is going on, on the CI Server.

**Comment by cbornet** [comment]
So the code generation is indeed not fully predictable. See the shippable log. Not sure what to do from there...

File: bin/ensure-up-to-date [hunk 1]
```diff
@@ -26,7 +26,9 @@ sleep 5
 ./bin/openapi3/php-petstore.sh
 
 # Check:
+echo "Perform git diff"
```
Comment (jmini) [inline bin/ensure-up-to-date]: I would have let this 3 lines (`echo`, `git diff` and `git status`) inside the if... We do not need them in the OK case.

**Comment by jmini** [comment]
Thank you for the addition.

Not having stable serialization is really bad. I guess `io.swagger.v3.core.jackson.SchemaSerializer` in swagger-parser should be improved. I will file an issue in Swagger-Parser.

---

In the mean time, I suggest that we remove spring from the list of checked samples. Remove this line:

https://github.com/OpenAPITools/openapi-generator/blob/7dfd940023a717649aa9af22100690ddab2731c1/bin/ensure-up-to-date#L15

I prefer the build to be green over additional checks.

**Comment by jmini** [comment]
I have proposed a way to make Yaml serialization deterministic with: #233. 

I performed some tests locally (I have merged #233 into the branch of this PR). I think it is a solution.

**Comment by jmini** [comment]
@cbornet: Now that #233 is merged I wanted to update this PR. In order to do so you need to grant access:

* "allow edits from maintainers" on the right side ([read more](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/))
* I also think that your repository [cbornet/openapi-generator](https://github.com/cbornet/openapi-generator/) should be public.

When I try to push to your branch `spring-api-first` I get:
> Repository `cbornet/openapi-generator' is disabled. Please ask the owner to check their account.

You can pick the commit from my [`pull_184`](https://github.com/jmini/openapi-generator/tree/pull_184) branch. Just rebase your `spring-api-first` on top of my `pull_184` branch.

**Comment by jmini** [comment]
Shippable build is now OK. Thank you @cbornet.

**Pull request merged**